src/suggestion-box.ts

```diff
@@ -0,0 +1,383 @@
+
+    @queryAssignedElements({ slot: "__private-item" })
+    private_items: Array<HTMLElement>;
+
+    @property()
+    prevItem: number = 0;
+
+    @state()
+    has_more: boolean = false;
+
+    @state()
+    index: number = -1;
+
+    @property()
+    size: number = 0;
+
+    @state()
+    show_empty: boolean = true;
+
+    @property()
+    result: Array<{ [key: string]: string }> = [];
+
+    @property()
+    highlight: string = "";
+
+    @property()
+    curr_index: number = 0;
+
+    @property()
+    input_el: HTMLElement | null = null;
+
+    @property()
+    filtered_list: Array<{ [key: string]: string }> = [];
+
+    @property()
+    prevIndex: number = 0;
+
+    @state()
+    show_container = false;
+
+    private _outClickListener = this.windowClickEvent.bind(this);
+    private input_temp: HTMLElement | null | string = null;
+
+    connectedCallback() {
+        super.connectedCallback();
+        this.addEventListener("mouseout", this._mouseOutEvent);
+        document.addEventListener("click", this._outClickListener);
+    }
+    disconnectedCallback() {
+        super.disconnectedCallback();
+        this.removeEventListener("mouseout", this._mouseOutEvent);
+        document.addEventListener("click", this._outClickListener);
+    }
+    render() {
+        return html`
+            <div style="position:relative">
+                <div
+                    part="sgn-input"
+                    class="inputbox"
+                    @focusin=${this._inputFocusEvent}
+                    @keydown=${this.keyDownInputEvent}
+                >
+                    <slot
+                        name="sgn-input"
+                        @input=${this.InputEvent}
+                        @click=${this.inputfield}
+                        ><input type="search"
+                    /></slot>
+                </div>
+                <slot name="sgn-item-template" style="display:none"></slot>
+                <div
+                    part="sgn-container"
+                    style="display:${this.show_container ? "block" : "none"};"
+                >
+                    <div style="display:${this.show_empty ? "block" : "none"}">
+                        <slot name="sgn-empty"></slot>
+                    </div>
+                    <div
+                        style="display:${!this.show_empty ? "block" : "none"};"
+                        @click=${this.saveInputEvent}
+                        @mouseover=${this.mouseOverEvent}
+                    >
+                        <slot name="__private-item"></slot>
+                    </div>
+                </div>
+            </div>
+        `;
+    }
+    private _indexOfTarget(event: any) {
+        let el = event.target;
+        while (el) {
+            if (el.slot == "__private-item") break;
+            el = el.parentElement;
+        }
+        return this.private_items.indexOf(el);
+    }
+    saveInputEvent(event: Event) {
+        if (this.result[this._indexOfTarget(event)]) {
```

Comment:
We can the result of this in a variable so that we can avoid calling _indexOfTarget two times.